File: src/graphs2/sssp.cpp
// Bellman-Ford（CLRS 圖 22.4，含負權）與 Dijkstra（CLRS 圖 22.6）
#include <iostream>
#include <queue>
#include <string>
#include <vector>

const int INF = 1e9;
struct Edge { int u, v, w; };

int main() {
    const std::string name = "stxyz";

    // ---- Bellman-Ford：|V|-1 輪全邊鬆弛＋第 |V| 輪驗負環 ----
    std::vector<Edge> E = {{0,1,6},{0,3,7},{1,2,5},{1,3,8},{1,4,-4},
                           {2,1,-2},{3,2,-3},{3,4,9},{4,2,7},{4,0,2}};
    std::vector<int> d(5, INF);
    d[0] = 0;
    for (int round = 1; round <= 4; ++round)          // |V| - 1 = 4 輪
        for (const Edge& e : E)
            if (d[e.u] != INF && d[e.u] + e.w < d[e.v])
                d[e.v] = d[e.u] + e.w;                // RELAX(u, v, w)
    bool ok = true;                                    // 第 |V| 輪：還能鬆 = 負環
    for (const Edge& e : E)
        if (d[e.u] != INF && d[e.u] + e.w < d[e.v]) ok = false;
    std::cout << "Bellman-Ford（源點 s）回傳 " << (ok ? "TRUE" : "FALSE") << "：";
    for (int v = 0; v < 5; ++v) std::cout << name[v] << ":" << d[v] << " ";
    std::cout << "\n";

    // ---- Dijkstra：非負權，priority_queue 貪婪取最近白點 ----
    std::vector<Edge> E2 = {{0,1,10},{0,3,5},{1,2,1},{1,3,2},{2,4,4},
                            {3,1,3},{3,2,9},{3,4,2},{4,0,7},{4,2,6}};
    std::vector<std::vector<std::pair<int,int>>> adj(5);
    for (const Edge& e : E2) adj[e.u].push_back({e.v, e.w});
    std::vector<int> dist(5, INF);
    using PII = std::pair<int,int>;                    // {dist, v}
    std::priority_queue<PII, std::vector<PII>, std::greater<PII>> Q;
    dist[0] = 0; Q.push({0, 0});
    while (!Q.empty()) {
        auto [du, u] = Q.top(); Q.pop();
        if (du > dist[u]) continue;                    // 過期項目跳過
        for (auto [v, w] : adj[u])
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w; Q.push({dist[v], v});
            }
    }
    std::cout << "Dijkstra（源點 s）：";
    for (int v = 0; v < 5; ++v) std::cout << name[v] << ":" << dist[v] << " ";
    std::cout << "\n";
    return 0;
}

File: src/graphs2/sssp.py
# Bellman-Ford＋Dijkstra（改寫自官方 Chapter 22，MIT License）：平鋪版
import heapq

INF = float("inf")

# ---- Bellman-Ford：CLRS 圖 22.4（含負權）----
E = [("s","t",6),("s","y",7),("t","x",5),("t","y",8),("t","z",-4),
     ("x","t",-2),("y","x",-3),("y","z",9),("z","x",7),("z","s",2)]
d = {v: INF for v in "stxyz"}
d["s"] = 0
for _ in range(4):                          # |V| − 1 輪全邊鬆弛
    for u, v, w in E:
        if d[u] + w < d[v]:
            d[v] = d[u] + w                 # RELAX(u, v, w)
ok = all(d[u] + w >= d[v] for u, v, w in E)  # 第 |V| 輪：還能鬆 = 負環
print("Bellman-Ford 回傳", ok, "：", " ".join("%s:%d" % (v, d[v]) for v in "stxyz"))

# ---- Dijkstra：CLRS 圖 22.6（非負權），heapq lazy deletion ----
adj = {"s": [("t",10),("y",5)], "t": [("x",1),("y",2)], "x": [("z",4)],
       "y": [("t",3),("x",9),("z",2)], "z": [("s",7),("x",6)]}
dist = {v: INF for v in adj}
dist["s"] = 0
h = [(0, "s")]
while h:
    du, u = heapq.heappop(h)
    if du > dist[u]:                        # 過期項目跳過
        continue
    for v, w in adj[u]:
        if dist[u] + w < dist[v]:
            dist[v] = dist[u] + w
            heapq.heappush(h, (dist[v], v))
print("Dijkstra：", " ".join("%s:%d" % (v, dist[v]) for v in "stxyz"))
